lessons/basics/documentation.md

```diff
@@ -0,0 +1,331 @@
+---
+layout: page
+title: Documentation
+category: basics
+order: 10
+lang: en
+---
+
+Documenting Elixir code.
+
+## Table of Contents.
+
+- [Annotation](#annotation)
+  - [Inline Documentation](#inline-documentation)
+  - [Documenting Modules](#documenting-modules)
+  - [Documenting Functions](#documenting-functions)
+- [ExDoc](#exdoc)
+  - [Installing](#installing)
+  - [Generating Documentation](#generating-documentation)
+- [Best Practice](#best-practice)
+
+
+## Annotation
+
+How much we comment and what makes quality documentation remains a contentious issue within the programming world. However, we can all agree that documentation is important for ourselves and those working with our codebase. 
+
+Elixir treats documentation as a *first-class citizen*, offering various functions to access and generate documentation for your projects. The Elixir core provides us with many different attributes to annotate a codebase. Let's look at 3 ways:
+
+  - `#`~ For inline documentation.
+  - `@moduledoc` ~ For module level documentation.
+  - `@doc` ~ For function level documentation.
+
+### Inline Documentation
+
+Probably the simplist way to comment your code is with inline comments. Similar to Ruby or Python, Elixir's inline comment is denoted with a `#`, frequently known as a *pound*, or a *hash* depending on where you are from in the world. 
+
+Take this Elixir Script (greeting.exs):
+
+```elixir
+# Outputs 'Hello, chum.' to the console.
+IO.puts "Hello, " <> "chum."
+```
+
+Elixir, when running this script will ignore everything from `#` to the end of the line, treating it as throw away data. It may add no value to the operation or performance of the script, however when it's not so obvious what is going happening a programmer should know from reading your comment. Be mindful not to abuse the single line comment! Littering a codebase could become an unwelcome nightmare for some. It is best used in moderation.
+
+### Documenting Modules
+
+The `@moduledoc` annotator allows for inline documentation at a module level. It typically sits just under the `defmodule` decalation at the top of a file. The below example shows a one line comment within the `@moduledoc` decorator.
+
+```elixir
+defmodule Greeter do
+  @moduledoc """
+
+  Provides a function `hello/1` to greet a human
+
+  """
+
+  def hello(name) do
+    "Hello, " <> name
+  end
+end
+```
+
+We (or others) can access this module documentation using the `h` helper function within IEx. 
+
+```bash
```

Comment:
For IEx blocks I use elixir syntax highlighting, disable the prompt line number, and skip the bash portion: 

```elixir
iex> c("greeter.ex")
[Greeter]
iex> h Greeter

                                    Greeter

Provides a function hello/1 to greet a human
```

Configuring the prompt is pretty straight forward:

```elixir
iex(2)> IEx.configure(default_prompt: "%prefix>")
:ok
iex> case :ok do
...> :ok -> "Hi"
...> _ -> "Other"
...> end
"Hi"
```